File: problem2.cpp
# Time Complexity : O(1) for all operations
# Space Complexity : O(n) where n is the number of elements in stack
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# This code uses 2 stacks: stack s1 that stores all values and s2 which holds the value of the smallest value seen upto this element. 

class MinStack {
    stack<int> s1;
    stack<int> s2;

public:
    MinStack() {
        
    }
    
    void push(int val) {
        s1.push(val);
        s2.push( (!s2.empty() && s2.top()<val) ? s2.top() : val);
    }
    
    void pop() {

        s2.pop();
        s1.pop();

    }
    
    int top() {
        return (s1.top());
    }
    
    int getMin() {
        return (s2.top());
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: Problem2.py
# Time Complexity :  O(1)  (when initilizing a vertical it will take O(hash1))
# Space Complexity :  O(self.size)

# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : Dynamically allocating the array instead of initiazing at the beginning

class MinStack:

    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)

    def pop(self) -> None:
        val = self.stack.pop()
        if val == self.min_stack[-1]:
            self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]
